src/back/hlsl/writer.rs

```diff
@@ -1441,6 +1566,180 @@ impl<'a, W: Write> Writer<'a, W> {
+                                // Texture always first parameter
+                                self.write_value_type(module, image_ty)?;
+                                // Mipmap is a second parameter if exists
+                                const MIP_LEVEL_PARAM: &str = "MipLevel";
+                                if let crate::ImageQuery::Size { level: Some(_) } = query {
+                                    write!(self.out, ", uint {}", MIP_LEVEL_PARAM)?;
+                                }
+                                writeln!(self.out, ")")?;
+
+                                // Write function body
+                                writeln!(self.out, "{{")?;
+                                const RETURN_VARIABLE_NAME: &str = "ret";
+
+                                use crate::ImageDimension as IDim;
+                                use crate::ImageQuery as Iq;
+                                // GetDimensions Overloaded Methods
+                                // https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-to-getdimensions#overloaded-methods
+                                let (ret_swizzle, number_of_params) = match query {
+                                    Iq::Size { .. } => match dim {
+                                        IDim::D1 => ("x", if arrayed { 2 } else { 1 }),
+                                        IDim::D2 => ("xy", if arrayed { 4 } else { 3 }),
+                                        IDim::D3 => ("xyz",  4),
+                                        IDim::Cube => ("xy", if arrayed { 4 } else { 3 }),
+                                    },
+                                    Iq::NumLevels | Iq::NumSamples | Iq::NumLayers => {
+                                        if arrayed || dim == IDim::D3 {
+                                            ("w", 4)
+                                        } else {
+                                            ("z", 3)
+                                        }
+                                    }
+                                };
+
+                                // Write `GetDimensions` function.
+                                writeln!(
+                                    self.out,
+                                    "{}uint4 {};",
+                                    back::INDENT,
+                                    RETURN_VARIABLE_NAME
+                                )?;
+                                write!(self.out, "{}", back::INDENT)?;
+                                self.write_expr(module, image, func_ctx)?;
+                                write!(self.out, ".GetDimensions(")?;
+                                match query {
+                                    Iq::Size { level: Some(_) } => {
+                                        write!(self.out, "{}, ", MIP_LEVEL_PARAM)?;
+                                    }
+                                    _ => match (dim, class) {
```

Comment:
this is a weird match! Effectively we are only checking each of those once. Is the idea to extend this later?
If we aren't going to touch this, let's opt into separate if lets or matches